[TP-13] [Blog] Blog page CTA block component

jhonatan-lopes opened the pull request:
# Description

<!-- Describe the PR here -->

Implement a CTA Block component.

Some changes were made from the original issue. I had to break this down into two components:

- A "CTA card" block, where we can display the required info
- A "CTA card with text" block, where we link a card to a "paragraph" so that the card can be placed either to the left or the right.

The "CTA card" block will display the "Full width" option. There is no option to add a "full width" card on the "CTA card with text" block, as this is equivalent to a regular "CTA card" block followed by a previously-available "Paragraph" block.

The `image`/`button` are optional, as well as a `title` for the card, which is styled differently from the rest of the text.

The `body` is a rich text field, allowing for:

- Bold
- Italics
- Links
- Unordered lists
- Ordered lists
- H4 and H5
- Horizontal rules

Link to sample test page: https://foundation-s-feature-87-h6m1io.mofostaging.net/en/blog/initial-test-blog-post-with-fixed-title/
Related PRs/issues: #8788 

# Checklist

<!-- Check off items with `[x]` or cross out items that don't apply with `~~The description~~` -->

**Tests**
- [X] Is the code I'm adding covered by tests? (PS.: only new custom blocks are covered, as blog page has no tests setup and its factory relies on the [old method that is going to be replaced](https://github.com/MozillaFoundation/foundation.mozilla.org/issues/11067))

**Changes in Models:**
- [ ] Did I update or add new fake data? -- Nope, see above. I've added factories, but not new fake data.
- [X] Did I squash my migration?
- [X] Are my changes [backward-compatible](https://github.com/mozilla/foundation.mozilla.org/blob/main/docs/workflow.md#django-migrations-what-to-do-when-working-on-backward-incompatible-migrations). If not, did I schedule a deploy with the rest of the team?

**Documentation:**
- ~~[ ] Is my code documented?~~
- ~~[ ] Did I update the READMEs or wagtail documentation?~~

**Merge Method**
**💡❗Remember to use squash merge when merging non-feature branches into `main`**

jhonatan-lopes commented:
Hey @nancyt1, the review app is ready now. I've added some cards to the CMS so that you can see the design

nancyt1 reviewed (changes requested):
Could we increase the padding on all sides by .5rem? 

Can we decrease the size of the top border to 4px?
<img width="715" alt="SCR-20230922-nxzq" src="https://github.com/MozillaFoundation/foundation.mozilla.org/assets/74557207/8ece8a5a-c884-47a6-813e-92bfc271f4a7">

Also decrease shadow of this box to 4px?
<img width="885" alt="SCR-20230922-nzae" src="https://github.com/MozillaFoundation/foundation.mozilla.org/assets/74557207/3a073743-ac3d-422a-b412-5353ec05f60a">

On the mockups, I originally wanted them to break out of the body gird, is this still possible?
<img width="924" alt="Screenshot 2023-10-02 at 1 37 31 PM" src="https://github.com/MozillaFoundation/foundation.mozilla.org/assets/74557207/2f3e0e5d-eea5-4055-9d21-31f077e1c26f">

jhonatan-lopes commented:
Hi @nancyt1, the review app is ready for you again

jhonatan-lopes commented:
Hi @nancyt1, I've addressed your comments, can you please take a look at this again?

More specifically:
* Increased the padding inside the card by 0.5 rem
* Decreased the shadows to 4px
* Implemented the break outside the grid card (finally able to get this one)

![image](https://github.com/MozillaFoundation/foundation.mozilla.org/assets/61277874/23c73d46-36dc-482a-bddc-4128590f209a)


To make the card break outside the grid, I'm using a negative margin of 12rem, please let me know if you would like to adjust this value.

I've only put this on cards that are aligned to the right. Cards aligned to left are then aligned to the text column.

I've also put a media query to make sure that this out of the grid effect only activates for large screens.

nancyt1 reviewed (changes requested):
It's looking great! I'm just trying to identify edge cases at this point. Let me know if there's anything I should elaborate on!

- Can we constrain images to fit the width of the card? Currently it makes the CTA huge if the image is big.
- Is it possible to also break out the left align one in a similar way as the right aligned one? Or maybe we can just remove left align option? Alternative is making left align smaller width so that it doesn’t take up more than 1/3 of the body columns.
- Do you have suggestions on how to prevent gaps like this if the paragraph text is shorter than the CTA text? Would adding character count be enough (~600-700)? 
<img width="1231" alt="SCR-20231006-phne" src="https://github.com/MozillaFoundation/foundation.mozilla.org/assets/74557207/8c436b0a-af10-4302-925f-356d4df38f32">



**For Mobile/Tablet:** 
- Can the cards be full width for mobile? Or at least be center aligned?
- For smaller screens like tablet, can the width of the card be ~20rem or take up no more than half of the body columns?
- Can we also reorder it so that the card comes after the paragraph text? Maybe we can also reorder in wagtail too.

**For Wagtail UI:** 
- Move the style and alignment options to the top
- Image and button should be optional, can we collapse them and let staff add it themselves?
- Can the title styling be H3? And can we add H5 to the styling options?

jhonatan-lopes commented:
Hi @nancyt1, thanks for the feedback!

I have altered the cards to also break from the grid when left aligned:
![image](https://github.com/MozillaFoundation/foundation.mozilla.org/assets/61277874/d28bf852-c9f9-458a-ac91-4f0a67598243)

I've put a validation that the paragraph content cannot be shorter than the card's content:
![image](https://github.com/MozillaFoundation/foundation.mozilla.org/assets/61277874/2da80b49-f6e7-47a0-9dfa-a0f47b1b80ea)


Full-width cards for mobile:
![image](https://github.com/MozillaFoundation/foundation.mozilla.org/assets/61277874/c994820b-3878-4519-87d0-a391382318ac)


For tablet, I've set it up to take 20rem max. I've also disabled the grid offset on tablet size:
![image](https://github.com/MozillaFoundation/foundation.mozilla.org/assets/61277874/4ef4b084-21f7-4bc8-acce-3f9e72e3fef1)

Cards should come after the paragraph on mobile screens now.

For Wagtail/UI:

- I've moved the alignment option to the top, but the style can only be at the top of the "card" section (so not at the top of the whole block, but at the top of the "card" sub-block)
- Image and button should be collapsed by default now
- I've changed the card title to be H3 and put options for H4 and H5 on the card body

The only thing left now is to address the image situation for full-width cards, as we discussed on slack.

jhonatan-lopes commented:
It was decided that the images for "full-width" cards should have their heights capped at `7rem`.

Here is the new review app:
https://foundation-s-feature-87-h6m1io.mofostaging.net/en/blog/initial-test-blog-post-with-fixed-title/

danielfmiranda commented on network-api/networkapi/wagtailpages/tests/customblocks/test_blog_cta_card_block.py:
```diff
@@ -0,0 +1,34 @@
+class TestBlogCTACardBlock(TestCase):
+    def test_blog_cta_card_all_optionals(self):
+        full_card = customblock_factories.BlogCTACardBlockFactory()
+        print(full_card)
```
Hey @jhonatan-lopes just checking if we want to print `full_card` during this test or if maybe this was left here by accident

jhonatan-lopes replied:
Great catch, thanks! I have a tendency to leave those pesky prints behind :see_no_evil:

danielfmiranda reviewed (approved):
Thanks @jhonatan-lopes! I think this PR sets a great example of creating factories for streamfield blocks.

Everything looks good to me. Approved!

nancyt1 reviewed (changes requested):
Sorry I only have 3 more notes!

1. Can we move the Alignment field back next to Styling? Sorry, it's more confusing now that it's by itself. 

2. Can we increase margin-right for the left-align block to 1.75rem? On tablet, can we also remove the margin-left indentation similar to the right-align block? <img width="1111" alt="SCR-20231017-mhng" src="https://github.com/MozillaFoundation/foundation.mozilla.org/assets/74557207/6e38a075-330b-4617-b7e3-2a87ebfa0cc2">

3. Can we update the full width CTA to also be full width for mobile? Currently they are slightly off from each other.
<img width="245" alt="Screenshot 2023-10-17 at 5 56 45 PM" src="https://github.com/MozillaFoundation/foundation.mozilla.org/assets/74557207/15c7837c-b84a-4465-9b62-232ab4f0ef7f">

jhonatan-lopes commented:
Hi @nancyt1, no problem! I've implemented the changes in your notes now

nancyt1 reviewed (approved):
Looks good, thank you!

mmmavis commented on network-api/networkapi/wagtailpages/templates/wagtailpages/blocks/blog_cta_card_with_text_block.html:
```diff
@@ -0,0 +1,14 @@
+
+{% block block_content %}
+    <div class="blog-cta-card-with-text">
+        <div class="blog-cta-card-with-text__card-container blog-cta-card-with-text__card-container--{{ value.alignment }}">
```
Hi @jhonatan-lopes , we don't use BEM naming convention in this codebase so I would appreciate if we don't  introduce a new naming pattern. To be honest, there's no one consistent naming convention that we have been sticking to in this codebase 😓 , however, we try to use shorter names possible.

I see that we are also adding new SCSS code in this PR. As we are trying to move away from SCSS, could we convert the rules into Tailwind classes instead? If the rules will be used more than just this block, we can even create custom component on Tailwind config. What do you think? If this sounds good then that also mitigate the naming convention headache here.

jhonatan-lopes replied:
Hi @mmmavis, thanks for raising these points! It was definitely inertia from working with BEM before :smile: I've migrated the styles to only use Tailwind now.

mmmavis reviewed (changes requested):
Hey @jhonatan-lopes sorry for getting to this PR late and thanks for the great work! I just have some questions regarding SCSS/Tailwind and class naming conventions. Please see my inline comments.

jhonatan-lopes commented on network-api/networkapi/wagtailpages/templates/wagtailpages/blocks/blog_cta_card_block_regular.html:
```diff
@@ -0,0 +1,32 @@
+{% load l10n wagtailcore_tags wagtailimages_tags %}
+
+<div class="[&_.rich-text_a]:tw-underline [&_.rich-text_a]:tw-text-black
```
I've tried to create a [variant plugin](https://tailwindcss.com/docs/plugins#static-variants) for this selector as we might need to use it often but it wouldn't work.

After adding

```js
    plugin(function({ addVariant }) {
      addVariant("rich-text-links", ["&_.rich-text_a"]);
    })
```
to our `tailwind.config.js` and changing the calls to `rich-text-links:tw-underline rich-text-links:tw-text-black`, the styles were no longer applied to links inside the cards.

This is a weird issue, since the syntax for the plugin seems to be correct, as we can see in this [Tailwind play environment](https://play.tailwindcss.com/eZmFo3HlUs).

jhonatan-lopes replied:
It seems like this is caused by a bug with Tailwind and prefixes: https://github.com/tailwindlabs/tailwindcss/issues/10125#issuecomment-1382724160

There is a workaround, which I have now implemented here :)

mmmavis commented on tailwind.config.js:
```diff
@@ -244,6 +245,10 @@ module.exports = {
         purple: "#353552",
       },
     },
+    boxShadow: {
```
@jhonatan-lopes Do you know if these "pop" and "outline" styles are also used elsewhere on the site? Or if they are defined in the design system? I'm a little worried that we add these styles as "boxShadow/pop/outline" to Tailwind configs because they are very specifically to this use case, and there are many other existing elements on the site have some sort of box shadow that we eventually need to make them adapt to Tailwind classes. If we want to keep these in the config, maybe rename these classnames to something less general could be an option?

jhonatan-lopes replied:
I see your point @mmmavis , thanks! I got the names from the design team, so I think these would be reused on the blog as it is defined with their design system. In the end, I thought this was "cleaner" than one time styles, but maybe I could rename them to be on a blog scope, e.g. "blog-pop" and "blog-outline"? What do you think?

jhonatan-lopes replied:
Here it is on the [design system](https://www.figma.com/file/528nyaJFzvCXAiw2gglwzh/Blog-Improvements?node-id=627%3A22439&mode=dev):
![image](https://github.com/MozillaFoundation/foundation.mozilla.org/assets/61277874/b99bc3d2-62c2-4ac7-af18-021101600b27)

mmmavis reviewed (changes requested):
Left a comment regarding Tailwind config. Everything else looks good!

mmmavis reviewed (approved):
Chatted with designers. It's good that we add these styles to Tailwind config.
